senseicore/src/channels.rs

```diff
@@ -0,0 +1,246 @@
+                    .iter()
+                    .enumerate()
+                    .find(|(_index, filter)| (filter.f)(event.clone()))
+                    .map(|(index, _filter)| index);
+
+                if let Some(index) = filter_index {
+                    events.push(event);
+                    filters.swap_remove(index);
+                }
+
+                if filters.is_empty() {
+                    return events;
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(interval_ms)).await;
+            current_ms += interval_ms;
+        }
+        events
+    }
+
+    pub async fn open_batch(
+        &mut self,
+        requests: Vec<ChannelOpenRequest>,
+    ) -> Vec<(ChannelOpenRequest, Result<[u8; 32], Error>)> {
+        let mut requests_with_results = requests
+            .into_iter()
+            .map(|request| {
+                let result = self.initiate_channel_open(&request);
+
+                (request, result)
+            })
+            .collect::<Vec<(ChannelOpenRequest, Result<[u8; 32], Error>)>>();
+
+        let filters = requests_with_results
+            .iter()
+            .filter(|(_request, result)| result.is_ok())
+            .map(|(request, _result)| {
+                let filter_node_id = self.node_id.clone();
+                let request_user_channel_id = request.custom_id;
+                let filter = move |event| {
+                    if let SenseiEvent::FundingGenerationReady {
+                        node_id,
+                        user_channel_id,
+                        ..
+                    } = event
+                    {
+                        if *node_id == filter_node_id && user_channel_id == request_user_channel_id
+                        {
+                            return true;
+                        }
+                    }
+                    false
+                };
+                EventFilter { f: filter }
+            })
+            .collect::<Vec<EventFilter<_>>>();
+
+        // TODO: is this appropriate timeout? maybe should accept as param
+        let events = self.wait_for_events(filters, 15000, 500).await;
+
+        // set error state for requests we didn't get an event for
+        let requests_with_results = requests_with_results
+            .drain(..)
+            .map(|(request, result)| {
+                if result.is_ok() {
+                    let event = events.iter().find(|event| {
+                        if let SenseiEvent::FundingGenerationReady {
+                            user_channel_id, ..
+                        } = event
+                        {
+                            if *user_channel_id == request.custom_id {
+                                return true;
+                            }
+                        }
+                        false
+                    });
+
+                    if event.is_none() {
+                        return (request, Err(Error::FundingGenerationNeverHappened));
```

Comment:
could the event happen later (after `wait_for_events`) if the peer is temporarily disconnected?